include/circt/Dialect/OM/OMOpInterfaces.td

```diff
@@ -15,6 +15,102 @@
 
 include "mlir/IR/OpBase.td"
 
+def ClassFieldsLike : OpInterface<"ClassFieldsLike"> {
+  let cppNamespace = "circt::om";
+
+  let description = [{
+    Marks the end of a class-like definition by providing a list of accessible
+    fields.
+  }];
+  let methods = [
+    InterfaceMethod<"Get the class-like field to type map",
+      "std::optional<mlir::Type>", "getFieldType", (ins "mlir::StringAttr":$name)>,
+    InterfaceMethod<"Print class fields like",
+      "void", "printClassFieldsLike", (ins "mlir::OpAsmPrinter &":$printer,
+                                           "llvm::SmallVector<mlir::Value>":$operands),
+      /*methodBody=*/[{}],
+      /*defaultImplementation=*/[{ 
+        // * assumes fieldNames dictionary attr exists
+        // * assumes each name in fieldNames has an entry in fieldTypes
+        //   dictionary attr mapping the name to type
+        // * perFieldAttrs, perFieldLocs optional attr/locs for each field
+        // * each field is emitted on a separate line
+        mlir::Operation *op = $_op.getOperation();
```

Comment:
IMHO this is too much C++ to write inline in a tablegen file. There aren't hard and fast rules, but if something gets to this level, and you still want to give a default implementation, I'd suggest defining a free function and calling it from the defaultImplementation block. As Hideto pointed out, maybe this doesn't need to be an interface method anyway, but just a thought about C++ in tablegen in general.